Add `$` tagged templates API

aaronccasanova opened the pull request:
Fixes #484 

This PR introduces a `$` API inspired by [google/zx](https://www.npmjs.com/package/zx), which allows users to execute commands using familiar `execa` options with the elegant DX of [tagged templates](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates):
 
#### Basic
```js
import {$} from 'execa';

const {stdout} = await $`echo unicorns`;
// const {stdout} = await $`echo ${'unicorns'}`;
// const {stdout} = await $`echo ${['unicorns', 'rainbows']}`;

console.log(stdout);
//=> 'unicorns'
```

#### With options
```js
import {$} from 'execa';

await $({stdio: 'inherit'})`echo unicorns`;
//=> 'unicorns'
```

#### With pre-defined options
```js
import {$} from 'execa';

const $$ = $({stdio: 'inherit'});

await $$`echo unicorns`;
//=> 'unicorns'
await $$({shell: true})`echo unicorns && echo rainbows`;
//=> 'unicorns'
//=> 'rainbows'
```

#### Synchronous
```js
import {$} from 'execa';

const {stdout} = $.sync`echo unicorns`;
console.log(stdout);
//=> 'unicorns'

$({stdio: 'inherit'}).sync`echo rainbows`;
//=> 'rainbows'
```

Co-authored-by: ehmicky <ehmicky@users.noreply.github.com>

ehmicky commented on index.js:
```diff
@@ -224,6 +224,43 @@ export function execaSync(file, args, options) {
+}
+
+function parseExpression(expression) {
+	return Array.isArray(expression) ? expression.join(' ') : expression ?? '';
```
The current logic would remove whitespace escaping. For example:

```js
$`file ${'Hello world'}` -> execa('file', ['Hello', 'world'])
```

instead of:

```js
$`file ${'Hello world'}` -> execa('file', ['Hello world'])
```

Also `parseCommand()` allow backslashes as escape characters, even though they are not needed there, so this function might not fit there. Calling just `string.split()` should be enough actually.

We also need to take into account that `template` might be `undefined`, when the template string contains invalid backlash sequences like an isolated `\u`. 

One possible implementation could be:

```js
const $ = (templates, ...expressions) => {
	const {file, args} = parseTemplates(templates, expressions);
	return execa(file, args);
};

const parseTemplates = (templates, expressions) => {
	const [file, ...args] = templates.flatMap((template, index) => parseTemplate(template, index, templates, expressions));
	return {file, args};
};

const parseTemplate = (template, index, templates, expressions) => {
	const templateString = template ?? templates.raw[index];
	const templateTokens = templateString.split(SPACES_REGEXP).filter(Boolean);

	if (index === expressions.length) {
		return templateTokens;
	}

	const expression = expressions[index];
	return Array.isArray(expression) ? [...templateTokens, ...expression.map(String)] : [...templateTokens, String(expression)];
};

const SPACES_REGEXP = / +/g;
```

I left the sync logic and the options binding logic out for the sake of the example.

This produces the following result:

```js
$`file one two` -> execa(file, ['one', 'two'])
$`file  one  two` -> execa(file, ['one', 'two'])
$`file ${'one'} ${'two'}` -> execa(file, ['one', 'two'])
$`file ${0}` -> execa(file, ['0'])
$`file "one two"` -> execa(file, ['"one', 'two"'])

$`file ${'one two'}` -> execa(file, ['one two'])
$`file ${['one', 'two three']}` -> execa(file, ['one', 'two three'])
$`file \n` -> execa(file, ['\n'])
$`file ${'\n'}` -> execa(file, ['\n'])
$`file \u` -> execa(file, ['\\u'])
$`file ${undefined}` -> execa(file, ['undefined'])
$`file one\\ two` -> execa(file, ['one\\', 'two'])
$`file ${'"one two"'}` -> execa(file, ['"one two"'])
```

Compared to the current implementation:

```js
$`file one two` -> execa(file, ['one', 'two'])
$`file  one  two` -> execa(file, ['one', 'two'])
$`file ${'one'} ${'two'}` -> execa(file, ['one', 'two'])
$`file ${0}` -> execa(file, ['0'])
$`file "one two"` -> execa(file, ['"one', 'two"'])

$`file ${'one two'}` -> execa(file, ['one', 'two'])
$`file ${['one', 'two three']}` -> execa(file, ['one', 'two', 'three'])
$`file \n` -> execa(file, [])
$`file ${'\n'}` -> execa(file, [])
$`file \u` -> execa(undefined, [])
$`file ${undefined}` -> execa(file, [])
$`file one\\ two` -> execa(file, ['one two'])
$`file ${'"one two"'}` -> execa(file, ['"one', 'two"'])
```

What do you think?

aaronccasanova replied:
Ah! Okay these examples are great and really drive home your point in our original discussion:

> Which means that quotes and backslashes do not need to be interpreted as escaping characters, since ${...} already fulfills escaping needs. They would be interpreted as regular characters instead.

I'm totally on board with this approach! Also, thanks for the sample implementation. It works great!
https://github.com/sindresorhus/execa/pull/510/commits/8aa4e0714cb98198641eed48b9aeb7266f6c4346

ehmicky commented:
Thanks a lot for tackling this and for navigating the back-and-forth of the discussion about escaping! :heart: 

Before digging too deep into the review, I left a general comment about escaping. Please let me know what you think. Thanks!

aaronccasanova commented on index.d.ts:
```diff
@@ -546,6 +546,64 @@ console.log(stdout);
+/**
+Same as `execa()` except both file and arguments are specified in a single `command` string. For example, `execa('echo', ['unicorns'])` is the same as $\`echo unicorns\`.
+
+If the file or an argument contains spaces, they must be escaped with backslashes. This matters especially if `command` is not a constant but a variable, for example with `__dirname` or `process.cwd()`. Except for spaces, no escaping/quoting is needed.
```
Note: I'll update this JSDoc once we iron out details on escaping

ehmicky commented on lib/command.js:
```diff
@@ -39,3 +39,22 @@ export const parseCommand = command => {
 	return tokens;
 };
+
+export const parseTemplate = (template, index, templates, expressions) => {
```
```suggestion
const parseTemplate = (template, index, templates, expressions) => {
```

ehmicky commented on index.js:
```diff
@@ -224,6 +224,30 @@ export function execaSync(file, args, options) {
+	return (templates, ...expressions) => {
+		const [file, ...args] = parseTemplates(templates, expressions);
+		return execa(file, args, templatesOrOptions);
+	};
```
Should we add a way to bind options multiple times? With each call being shallowly merged to the previous one. This could be useful when someone wants to set some global options once, but then add more options for specific calls. 

Also, when users might be binding options before knowing whether the call will be sync or not. Or they might bind options for multiple calls, sync and async.

For example:

```js
// Global options
$ = $({ env: { ... } })

// ...

// Additional options for a specific call
const result = await $({ shell: true })`ls *`
// And also a sync call
await $.sync({ cwd: '...' })`rm foo`
```

What are your thoughts on this?

aaronccasanova replied:
While this behavior could be well documented, my initial thoughts are that this could also be confusing for users and difficult to provide accurate type definitions.

A similar experience, albeit less elegant, might be encouraging the following workflows:

1. Passing around a global `execaOptions` object
```js
// src/constants.js
export const execaOptions = {...}

// src/script.js
import {execaOptions} from './constants.js'

await $({...execaOptions, shell: true})`ls *`
```

2. Create a project specific wrapper for the `$` API
```js
// src/$.js
import {$ as base$} from 'execa';

const execaOptions = {...}

export const $ = (templatesOrOptions, ...expressions) => {
  if (Array.isArray(templatesOrOptions)) {
    return base$(execaOptions)(templateOrOptions, ...expressions);
  }

  return (templates, ...expressions) => {
    return base$({...execaOptions, ...templatesOrOptions})(templates, ...expressions);
  };
}

// src/script.js
import {$} from './$.js'

await $({shell: true})`ls *`
```

> Note: We could also expose the above recipe as a `create$(baseOptions)` utility

In both cases, I think the code might be easier for users to understand and navigate within their projects. Thoughts?

ehmicky replied:
I think the second option might be too much work for users. However, the first option is a possibility.

Yet another possibility would be to use an imperative approach instead of functional, i.e. expose a global mutable `$.options` object that users can modify. I personally dislike that kind of approach because it unnecessarily introduces global state. However, it might be easy to understand and avoid passing option objects around.

I'd like to get @sindresorhus input on this before moving forward. @sindresorhus What do you think?

Summary: how to pass options to `$` tagged template?
  1. `$(options)` binds options. For global options, it can be called several times.
  2. `$(options)` binds options, but can be called only once. For global options, users are expected to pass options objects around.
  3. `$.options` global object.

ehmicky replied:
@sindresorhus What do you think?

sindresorhus replied:
It should definitely be immutable. Options should be passed in normally. Would also be nice to provide an easy API to create a new `$` with custom options (as shown with the "Above example with create$ API" suggestion above). So `2` with a convenience API to create new instances bound to some options.

ehmicky replied:
This sounds good, and take the best of both approaches.
What about renaming the following?

```js
// Global options
$ = $({ env: { ... } })
```

To:

```js
// Global options
$ = $.create({ env: { ... } })
```

This removes the need for two different methods `$(options)` and `$.create(options)` which would be quite similar to each other. Like this, users only need to learn a single method `$.create(options)`.

sindresorhus replied:
Agreed

ehmicky replied:
@aaronccasanova What do you think?

aaronccasanova replied:
I like it! I'll do a quick experiment with the `$.create` API and let ya'll know if I hit any blockers (already anticipating some TypeScript complexities).

Some initial thoughts, concerns, clarifications:

**concern:** The examples thus far have been re-assigning the `$` API (e.g. `$ = $.create(...)`). However, from my understanding, this won't really be possible without renaming the imported `$`:
<details>
  <summary>Can't re-assign constant variable <code>$</code></summary>

<img width="904" alt="Screen Shot 2023-02-12 at 10 47 06 AM" src="https://user-images.githubusercontent.com/32409546/218331739-486922f8-8b0c-40ea-b397-2d4d83ced4bc.png">
</details>
<details>
  <summary>Can't import <code>$</code> variable with same name as local identifier</summary>

<img width="899" alt="Screen Shot 2023-02-12 at 11 01 43 AM" src="https://user-images.githubusercontent.com/32409546/218331923-a3cebfce-ac45-4c02-bf0a-1cca1af2d07d.png">
</details>

Perhaps, this is not an issue, but a bit clunky IMO

Note: The `create$` API doesn't have this "issue", but removes the ability to recursively assign and merge options e.g. `$ = $.create(...) = $.create(...) = $.create(...)`

**clarification:** Just want to make sure my previous example is inline with the intention of the API (e.g. `$.create` can be recursively re-assigned and options are shallow merged). I assume so from @sindresorhus's comment: `So each call can extend the existing defaults. ``.create`` can make new defaults be merging with existing ones.`

ehmicky replied:
Good catch, you're correct.
So users would need to use a different name and variable than `$` for the return value of `$.create()`. I am actually thinking this might be a good thing? Like this, the original `$` is immutable, and options-bound "copies" of `$` have their own distinct variables, which might create less confusion for users.

```js
import { $ } from 'execa'

const $$ = $.create({ env: { ... } })
await $$({ shell: true })`ls *`
await $$.async({ cwd: '...' })`rm foo`
```

Or:

```js
import { $ as $$ } from 'execa'

const $ = $$.create({ env: { ... } })
await $({ shell: true })`ls *`
await $.async({ cwd: '...' })`rm foo`
```

sindresorhus replied:
> I am actually thinking this might be a good thing? Like this, the original $ is immutable, and options-bound "copies" of $ have their own distinct variables, which might create less confusion for users.

👍

aaronccasanova replied:
Sounds good 👍 Okay, here's my initial pass at the `$.create` API: https://github.com/sindresorhus/execa/pull/510/commits/5fac1b7ecdafa4b871a81aaeb191cf537116c4ca

I'll pause here for feedback before moving on to types, tests, and docs!

aaronccasanova replied:
While addressing the eslint errors, I noticed and included refactor that removed all the duplicate conditional logic: https://github.com/sindresorhus/execa/pull/510/commits/1bd04ec27d5a055671e7bd8e0983dcde6fd0939e

ehmicky replied:
Thanks @aaronccasanova,

The current implementation allows the following two syntaxes:

```js
await $({ shell: true })`ls *`
await $.create({ shell: true })`ls *`
```

I suggested this above, but now I am wondering: those two are actually functionally equivalent, so would it make sense to merge `$(...)` and `$.create(...)` into a single method?

aaronccasanova replied:
I considered this as well, but still liked the idea of a dedicated method for binding options. To play devils advocate, if we merged `$(...)` and `$.create(...)` are there any concerns around creating extraneous functions and memory? Here's a super contrived example:

```js
const $$ = $({ shell: true })

for (const command of commands) {
  await $$(command.options)(...command.args)
}
```

In this example, each `$$(...)` would create a brand new `$` and `$.sync` function. As opposed to calling the same reference returned by `$.create(...)`.

ehmicky replied:
When it comes to memory consumption, I think the memory and CPU time wasted by creating anonymous function is quite negligible compared to spawning the child process. For example, CPU time is in the milliseconds-range for child process spawning, while creating anonymous function is in the nanoseconds-range. Memory consumption is also along those lines.

I think having a dedicated method for options binding is nice too, but I am wondering whether there is a technical difference between: 
  - Binding options early (e.g. global options), then spawn multiple processes with it
  - Spawning a single process with specific options

Conceptually, those two could be done with a single method like `$(options)`. Having a single method instead of two could simplify the API for users, and also its implementation. :thinking:

ehmicky replied:
Great idea to move `$.sync(...)` to `$(...).sync`! I am wondering how @sindresorhus feels about this, but this look good to me!

aaronccasanova replied:
Okay, here are the two APIs/commits to compare:
- https://github.com/sindresorhus/execa/pull/510/commits/1bd04ec27d5a055671e7bd8e0983dcde6fd0939e: Implements `$.create` for binding options to both the async `$` and `$.sync` APIs and allows one off override options per call e.g. `$(overrideOptions)` and `$.sync(overrideOptions)`
- https://github.com/sindresorhus/execa/pull/510/commits/50055a6a521b5edb8cd805b907e13851a97dbc77: Consolidates option binding and merging to a single method e.g. `$(options)` and `$(options).sync`

sindresorhus replied:
> Having a single method instead of two could simplify the API for users, and also its documentation, implementation, typing, tests, etc. 

:+1:

> Great idea to move $.sync(...) to $(...).sync! I am wondering how @sindresorhus feels about this, but this look good to me!

:+1:

ehmicky replied:
The approach in the current state of the PR (50055a6) looks good to me, thanks again for working on this! :+1:

ehmicky commented on test/command.js:
```diff
@@ -85,3 +85,88 @@ test('execaCommandSync()', t => {
+});
+
+test('$ accepts options', async t => {
+	const {stdout} = await $({encoding: 'utf8'})`node test/fixtures/echo.js foo bar`;
```
We might want to use an example where the options changes `stdout`. The `encoding` does not change `stdout` here, so the test would pass even if the `encoding` option was ignored.

aaronccasanova replied:
Nice catch! For now, I just updated the tests to [properly validate the APIs accept options](https://github.com/sindresorhus/execa/pull/510/commits/0cc0fc3aa6072763252ebd664f7abfefe5e6a023), but agree we should include tests to validate different `stdout`s 👍

ehmicky commented on playground.js:
```diff
@@ -0,0 +1,12 @@
+$$ = $$({shell: true, env: {TEST: 'test'}});
+
+// eslint-disable-next-line no-unused-expressions
+$$.sync`echo "$TEST"`;
```
I am wondering whether you intended to commit this file, or whether it was meant for debugging instead?

aaronccasanova replied:
Thanks for flagging. Yeah, this file was temporarily committed for debugging and exploring the different options binding APIs we discussed above. Removed: https://github.com/sindresorhus/execa/pull/510/commits/3f6de3d730bbbe3ce54a426f6bef8213d74cfea3

ehmicky commented on index.d.ts:
````diff
@@ -546,6 +546,64 @@ console.log(stdout);
+//=> 'rainbows'
+```
+*/
+export const $: {
````
We should add some type tests as well.

aaronccasanova replied:
Added! https://github.com/sindresorhus/execa/pull/510/commits/a287fc97c2c60ada452bc1b0b3755e596623bfac

Note: I just copied the `execaCommand` and `execaCommandSync` tests and swapped out implementations with the respective `$` APIs. Thoughts?

ehmicky replied:
This makes sense. Thanks!

ehmicky commented on index.d.ts:
```diff
@@ -546,6 +546,64 @@ console.log(stdout);
 export function execaCommand(command: string, options?: Options): ExecaChildProcess;
 export function execaCommand(command: string, options?: Options<null>): ExecaChildProcess<Buffer>;
 
+type Expression = string | number | Array<string | number>;
```
Small nit-picking: what do you think about naming this type more specifically. `Expression` is rather generic, and I did not realize at first it was meant only for the template strings expressions under `${...}`. Maybe `TemplateExpression`, or anything else more specific?

aaronccasanova replied:
Good call! `TemplateExpression` is 👌 https://github.com/sindresorhus/execa/pull/510/commits/6ed5b89f7d37f15e707bed0239788ef2555e9f45

aaronccasanova commented on index.d.ts:
```diff
@@ -546,6 +546,64 @@ console.log(stdout);
+
+	@returns A result `Object` with `stdout` and `stderr` properties.
+	*/
+	sync<T extends TemplateStringsArray | Options<string | null>>(
```
It's definitely worth exploring 👍 I just pushed up some [initial type updates](https://github.com/sindresorhus/execa/pull/510/commits/e1365af02748a49aa48e0d323e9c14fcc2c1fcfd), matching the new options binding API (using generics), so we at least have a working solution. However, I can circle back to see if function overloads is an option when I start fleshing out the `index.test-d.ts` file.

aaronccasanova replied:
Wow, okay so using overloads drastically simplified the types imo. Thoughs? https://github.com/sindresorhus/execa/pull/510/commits/6a283cb0dd156c4060b3935f253be758dfceba98

ehmicky replied:
Yes, it did! Overloads can be quite nice in situations like this. :)

ehmicky commented on package.json:
```diff
@@ -61,6 +61,7 @@
 		"path-key": "^4.0.0",
 		"tempfile": "^4.0.0",
 		"tsd": "^0.19.1",
+		"type-fest": "^3.5.7",
```
I believe this should be a production dependency? Otherwise, TypeScript users that do `npm install execa` would receive a type error due to the missing dependency.

`type-fest` is rather big, so I would be wondering how @sindresorhus feels about introducing that dependency :thinking:

aaronccasanova replied:
I have since removed `type-fest` with the refactor to use type overloads: https://github.com/sindresorhus/execa/pull/510/commits/6a283cb0dd156c4060b3935f253be758dfceba98

ehmicky commented on test/command.js:
```diff
@@ -85,3 +85,88 @@ test('execaCommandSync()', t => {
+});
+
+test('$ accepts options', async t => {
+	const {stdout} = await $({shell: true, env: {TEST: 'test'}})`echo $TEST`;
```
I believe this might not work on `cmd.exe`? This is the shell used by `shell: true` on Windows, and environment variables are using [a different syntax `%TEST%`](https://github.com/ehmicky/cross-platform-node-guide/blob/main/docs/4_terminal/environment_variables.md#referencing).

What about using the `stripFinalNewline` option instead, and checking to see if the newline is stripped?

aaronccasanova replied:
Another great catch! Updated the tests to use the `stripFinalNewline` option: https://github.com/sindresorhus/execa/pull/510/commits/bddf3179e26a50aff59c9eec72d703bf55d9eb3a

ehmicky commented on index.test-d.ts:
```diff
@@ -219,3 +220,16 @@ expectType<ExecaChildProcess<Buffer>>(
+expectType<ExecaSyncReturnValue>($({encoding: 'utf8'}).sync`unicorns`);
+expectType<ExecaSyncReturnValue<Buffer>>($({encoding: null}).sync`unicorns`);
+expectType<ExecaSyncReturnValue>($({encoding: 'utf8'}).sync`unicorns foo`);
+expectType<ExecaSyncReturnValue<Buffer>>($({encoding: null}).sync`unicorns foo`);
```
We should add some tests with `${...}` to test the type of `TemplateExpression`. With `string`, `number`, `string[]` and `number[]`, and also with another type (like `boolean`), where it should then fail. What do you think?

aaronccasanova replied:
I'm not super familiar with `tsd`. How does this look? https://github.com/sindresorhus/execa/pull/510/commits/656a69b05c9bae4330503f295a8d6d8386ce3fa2

ehmicky replied:
That's exactly it. Thank!

ehmicky commented on index.d.ts:
```diff
@@ -541,6 +541,27 @@ console.log(stdout);
+type TemplateExpression = string | number | Array<string | number>;
+
+type Execa$<EncodingType = string> = {
+	(options: Options): Execa$;
```
```suggestion
	(options: Options): Execa$<EncodingType>;
```

We should also add some type tests for it, like:

```ts
expectType<ExecaReturnValue<Buffer>>(await $({encoding: null})({})`unicorns foo`);
expectType<ExecaReturnValue>(await $({encoding: null})({encoding: 'utf8'})`unicorns foo`);
```

aaronccasanova replied:
Included the above suggestion here: https://github.com/sindresorhus/execa/pull/510#discussion_r1109290963

aaronccasanova replied:
Included the above suggestion here: https://github.com/sindresorhus/execa/pull/510#discussion_r1109290963

ehmicky commented on index.d.ts:
```diff
@@ -541,6 +541,27 @@ console.log(stdout);
+	): EncodingType extends null
+		? ExecaSyncReturnValue<Buffer>
+		: ExecaSyncReturnValue;
+};
```
I believe the conditionals might be simplified by using directly:
 
```ts
type Execa$<StdoutErrorType = string> = {
	(options: Options): Execa$<StdoutErrorType>;
	(options: Options<null>): Execa$<Buffer>;
	(
		templates: TemplateStringsArray,
		...expressions: TemplateExpression[]
	): ExecaChildProcess<StdoutErrorType>;
	sync(
		templates: TemplateStringsArray,
		...expressions: TemplateExpression[]
	): ExecaSyncReturnValue<StdoutErrorType>;
};
```

What do you think?

aaronccasanova replied:
It just keeps getting better with each iteration! I'll update to this and add the above mentioned test cases 👌

aaronccasanova replied:
Actually, I'm seeing a mix of `StdoutStderrorType` and `StdoutErrorType` arguments thorought the file. I have a slight preference for the more explicit `StdoutStderrorType` variant. Mind if I standardize all references?

ehmicky replied:
Yes, absolutely, good catch :+1:

aaronccasanova replied:
I had to make one small adjustment, but your suggestions worked great! https://github.com/sindresorhus/execa/pull/510/commits/989f6bc1fa5f007106a9cc5a24dfd1e0e27ac033

> Yes, absolutely, good catch 👍

Done https://github.com/sindresorhus/execa/pull/510/commits/f40759b2cfeb6eeaf9fb5e7e302a655e3bb5c8fe

aaronccasanova commented:
Thanks for the incredibly thorough review and assistance with the implementation @ehmicky! I too am very happy with the API we've landed on and believe I have addressed all the feedback (with the exception of updating the `readme.md`).

Regarding the `readme.md`, what are we thinking here? Should we add a collection of [usage examples here](https://github.com/sindresorhus/execa#usage) or just add a `$` section to the [API docs](https://github.com/sindresorhus/execa#api)?

ehmicky commented:
Personally, I would do both. This is quite a big feature, so it'd deserve some higher visibility by showing a high-level usage example. And then, the API section would get into more details. What do you think?

aaronccasanova commented:
Sounds good 👍 I kicked off the process by copy/pasting the examples in the PR description, adjusted them to match other examples, and fixed up a few typos in the JSDoc types: https://github.com/sindresorhus/execa/pull/510/commits/f8fdd69ef2883b55c946066a91bebaee09029344

From here, I may need some assistance moving things around and dialing in the right balance between usage examples and written documentation.

ehmicky commented on readme.md:
```diff
@@ -191,6 +240,18 @@ Execute a file synchronously.
 
+### \$(options?)\`command\`
+
+Same as `execa()` except both file and arguments are specified in a single tagged template string. For example, ``$`echo unicorns` ``is the same as `execa('echo', ['unicorns'])`.
```
```suggestion
Same as [`execa()`](#execafile-arguments-options) except both file and arguments are specified in a single tagged template string. For example, ``$`echo unicorns` ``is the same as `execa('echo', ['unicorns'])`.
```

ehmicky replied:
We should have some information that any string or array of strings (not sure whether number arguments need be documented?) in `${...}` is automatically escaped.

ehmicky commented on readme.md:
```diff
@@ -191,6 +240,18 @@ Execute a file synchronously.
+
+Same as `execa()` except both file and arguments are specified in a single tagged template string. For example, ``$`echo unicorns` ``is the same as `execa('echo', ['unicorns'])`.
+
+The [`shell` option](#shell) must be used if the `command` uses shell-specific features (for example, `&&` or `||`), as opposed to being a simple `file` followed by its `arguments`.
```
We might want to add that quotes, backslashes and spaces are automatically escaped and have no special meaning unless the `shell` option is used. I can see this becoming a common mistake. About half of the issues in this repository are related to confusion about shell syntax and escaping.

ehmicky commented on readme.md:
```diff
@@ -191,6 +240,18 @@ Execute a file synchronously.
+
+Same as [$(options?)\`command\`](#optionscommand) but synchronous.
+
+Returns or throws a [`childProcessResult`](#childProcessResult).
```
The async method should include a similar line IMHO. This is implied by "Same as `execa()`". However, it might be good to be explicit about the return value. What do you think?

aaronccasanova replied:
Addressed here: https://github.com/sindresorhus/execa/pull/510#discussion_r1111455947

ehmicky commented on readme.md:
```diff
@@ -191,6 +240,18 @@ Execute a file synchronously.
+
+### $(options?).sync\`command\`
+
+Same as [$(options?)\`command\`](#optionscommand) but synchronous.
```
```suggestion
Same as [$(options?)\`command\`](#optionscommand) but synchronous like [`execaSync()`](#execasyncfile-arguments-options).
```

ehmicky commented on readme.md:
````diff
@@ -39,6 +39,55 @@ console.log(stdout);
+$({stdio: 'inherit'}).sync`echo rainbows`;
+//=> 'rainbows'
+```
+
````
Those examples are looking good to me, thanks! :tada:

ehmicky commented on index.d.ts:
```diff
@@ -541,8 +541,80 @@ console.log(stdout);
+
 /**
-Same as `execaCommand()` but synchronous.
+Same as `$` but synchronous.
```
This documents `execaCommandSync()`, so comparing with `execaCommand()` was correct?

aaronccasanova replied:
Whoops, that was unintentional: https://github.com/sindresorhus/execa/pull/510/commits/97bdf47bc93067322e51f01b2df7bd6758397b53

ehmicky commented on readme.md:
```diff
@@ -191,6 +242,24 @@ Execute a file synchronously.
+
+The return value of this function depends on whether you are binding options or executing a command:
+- If you are binding options, for example with `const $$ = $(options)`, the return value is a new `$` instance with the options bound to both the asynchronous `$` and synchronous `$.sync` APIs. It's important to note that consecutive calls to this API will shallow merge the options.
+- If you are executing a command, the result is a `Promise` that resolves or rejects with a [`childProcessResult`](#childProcessResult).
```
I am wondering whether we should split this section into two?
  - $\`command\`, which executes a command
  - `$(options)`, which returns a new `$` with specific default `options`

This might also make it clearer that users do not need to call `$(options)` to execute a command.

aaronccasanova replied:
Good call 👍 Here's my first pass at breaking the `$`, `$.sync`, and `$(options)` APIs into dedicated sections: https://github.com/sindresorhus/execa/pull/510/commits/33f4d065a0d02172ae23e658ce05126d7a679e89

sindresorhus commented on index.d.ts:
```diff
@@ -541,6 +541,78 @@ console.log(stdout);
+
+type Execa$<StdoutStderrType = string> = {
+	/**
+  Same as `execa()` except both file and arguments are specified in a single tagged template string. For example, `` $`echo unicorns` `` is the same as `execa('echo', ['unicorns'])`.
```
Incorrect indentation.

aaronccasanova replied:
Fixed https://github.com/sindresorhus/execa/pull/510/commits/38378d01e16a4258d34728a2d1cd1e96eaaf95cf

sindresorhus commented on index.d.ts:
```diff
@@ -541,6 +541,85 @@ console.log(stdout);
+	/**
+	Same as `$` but synchronous.
+
+	@returns A result `Object` with `stdout` and `stderr` properties.
```
The types already describe this. It could instead say:

```suggestion
	@returns The stdout and stderr output.
```

ehmicky commented on readme.md:
```diff
@@ -191,6 +242,28 @@ Execute a file synchronously.
+
+### $(options)
+
+Binds options to the [`$`](#command) API. For example, you can use `$(options)` to create a new `$` instance with specific default options, which are then bound to both the asynchronous `` $`command` `` and synchronous `` $.sync`command` `` APIs.
```
```suggestion
Binds options to the [`$`](#command) API. For example, you can use `$(options)` to create a new `$` instance with specific default options, which are then bound to both the asynchronous [`` $`command` ``](#command) and synchronous [`` $.sync`command` ``](#synccommand) APIs.
```

ehmicky commented on index.d.ts:
```diff
@@ -541,6 +541,87 @@ console.log(stdout);
+	): ExecaChildProcess<StdoutStderrType>;
+
+	/**
+	Same as `$` but synchronous.
```
The description of both `$` and `$.sync` in the types documentation should be a copy of the one in the `readme.md`.

aaronccasanova replied:
Done https://github.com/sindresorhus/execa/pull/510/commits/9e566e582e065529673b073c2dc2ed7e7e1e5741

sindresorhus commented on readme.md:
````diff
@@ -39,6 +39,57 @@ console.log(stdout);
+
+$({stdio: 'inherit'}).sync`echo rainbows`;
+//=> 'rainbows'
+```
````
The examples here do not exactly match the ones in the types. I think we should try to make them as similar as possible to ease maintenance.

https://github.com/sindresorhus/execa/pull/510/files#diff-7aa4473ede4abd9ec099e87fec67fd57afafaf39e05d493ab4533acc38547eb8R608-R625

aaronccasanova replied:
Good call 👍 I think the examples provided in the readme are comprehensive and demonstrate the full API, so I copied those over to the JSDoc types: https://github.com/sindresorhus/execa/pull/510/commits/5bb310ec9a9065265be381eeeab399f2f54ff995

sindresorhus commented:
Looks great! Thanks so much for working on this, @aaronccasanova 🙏

aaronccasanova commented:
Of course! Thanks for all your feedback and assistance with the API design, @sindresorhus! And a big shoutout to @ehmicky for the thorough review and support throughout the process 🙏

sindresorhus commented:
Just FYI. We will hold off from releasing this for another week or so as @ehmicky is writing a blog post about it.
